### PR TITLE
Fix freemarker rendering validation in 🧅 archi

### PR DESCRIPTION
**Pull request by jourdiw**
## Issue

https://gravitee.atlassian.net/browse/APIM-3955

## Description

When validating if an API has a valid template, the 🧅 architecture was passing a non-ApiModel object to the template engine. This resulted in validation errors for dates, differently named attributes, and missing information that exists in Portal rendering (such as `metadata` or `primaryOwner`).

Publishing to portal via MAPI-v2 is now possible:

Screenshot 2024-02-29 at 15 27 02

Template text:
```
<#if api.picture??>
<img src="${api.picture}" style="float: right;max-width: 60px;"/>
</#if>

# Welcome to the API ${api.name}(${api.version})!

The API is a ${api.definitionVersion} API. :)

The API is <span style="text-transform: lowercase;color: <#if api.state=='STARTED'>green<#else>red</#if>">${api.state}</span>.

This API has been created on ${api.createdAt?datetime} and updated on ${api.updatedAt?datetime}.

<#if api.deployedAt??> 
This API has been deployed on ${api.deployedAt?datetime}.
<#else>
This API has not yet been deployed. 
</#if>

<#if api.visibility=='PUBLIC'>
This API is publicly exposed.
<#else>
This API is not publicly exposed.
</#if>

<#if api.tags?has_content>
Sharding tags: ${api.tags?join(", ")}
</#if>

## Description

${api.description}

<#if api.proxy??>
## How to access

The API can be accessed through https://api.company.com${api.proxy.contextPath}:

curl https://api.company.com${api.proxy.contextPath}
</#if>

## Rating
    sssss

You can rate and put a comment for this API <a href='/#!/apis/${api.id}/ratings'>here</a>.

## Contact

<#if api.metadata['email-support']??>The support contact is <a href="mailto:${api.metadata['email-support']}">${api.metadata['email-support']}</a>.</#if>

The API owner is <#if api.primaryOwner.email??><a href="mailto:${api.primaryOwner.email}">${api.primaryOwner.displayName}</a><#else>${api.primaryOwner.displayName}</#if>.
```

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-qzyegmarty.chromatic.com)

### CODE REVIEW

**Review by a-cordier — APPROVED** [review]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/api/model/ApiFreemarkerTemplate.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import io.gravitee.definition.model.v4.ApiType;
+import io.gravitee.definition.model.v4.property.Property;
+import io.gravitee.definition.model.v4.resource.Resource;
+import java.util.*;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+public class ApiFreemarkerTemplate {
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/api/model/ApiFreemarkerTemplate.java]: Maybe we can move that file in `documentation` package as it is related to Documentation feature from what I understood 🤔

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/documentation/domain_service/DocumentationValidationDomainService.java [hunk 2]
```diff
@@ -50,8 +57,18 @@ public void validateContentIsSafe(String content) {
         }
     }
 
-    public void validateTemplate(String pageContent, String apiId) {
-        this.templateResolverDomainService.resolveTemplate(pageContent, Map.of("api", this.apiCrudService.get(apiId)));
+    public void validateTemplate(String pageContent, String apiId, String organizationId) {
+        var metadata = new HashMap<String, String>();
+        this.apiMetadataQueryService.findApiMetadata(apiId)
+            .forEach((k, v) -> metadata.put(k, v.getValue() != null ? v.getValue() : v.getDefaultValue()));
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/documentation/domain_service/DocumentationValidationDomainService.java]: I would have used a more functional way to built it by using a `.map` with maybe a `.collect` but 🤷

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/documentation/domain_service/TemplateResolverDomainService.java [hunk 3]
```diff
@@ -15,6 +15,7 @@
  */
 package io.gravitee.apim.core.documentation.domain_service;
 
+import io.gravitee.apim.core.api.model.Api;
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/documentation/domain_service/TemplateResolverDomainService.java]: 🗑️

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/domain_service/documentation/FreemarkerTemplateResolver.java [hunk 4]
```diff
@@ -20,8 +20,10 @@
 import freemarker.template.Configuration;
 import freemarker.template.Template;
 import freemarker.template.TemplateException;
+import io.gravitee.apim.core.api.model.Api;
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/domain_service/documentation/FreemarkerTemplateResolver.java]: 🗑️

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/domain_service/documentation/FreemarkerTemplateResolver.java [hunk 5]
```diff
@@ -20,8 +20,10 @@
...
 import io.gravitee.apim.core.documentation.domain_service.TemplateResolverDomainService;
 import io.gravitee.apim.core.documentation.exception.InvalidPageContentException;
+import io.gravitee.apim.infra.adapter.ApiAdapter;
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/domain_service/documentation/FreemarkerTemplateResolver.java]: 🗑️

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/inmemory/NoopTemplateResolverDomainService.java [hunk 6]
```diff
@@ -15,6 +15,7 @@
  */
 package inmemory;
 
+import io.gravitee.apim.core.api.model.Api;
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/inmemory/NoopTemplateResolverDomainService.java]: 🗑️

**Review by jgiovaresco — COMMENTED** [review]
I would have use the `FreemarkerTemplateResolver` in the use cases tests to add a test case checking that
- no error is thrown when the template is OK
- an error is thrown when the template contains inaccessible variable.

I would add these test in Use Case tests so. that when we want to understand the business rules of that use case we can just look at the tests

**Comment by jourdiw** [comment]
I will create a separate PR to create more use-cases for the Create/Update use-case test suites.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/adapter/ApiAdapter.java [hunk 7]
```diff
@@ -17,12 +17,12 @@
 
 import io.gravitee.apim.core.api.model.Api;
 import io.gravitee.apim.core.api.model.crd.ApiCRD;
+import io.gravitee.apim.core.documentation.model.ApiFreemarkerTemplate;
+import io.gravitee.apim.core.membership.model.PrimaryOwnerEntity;
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/adapter/ApiAdapter.java]: 🗑️

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/adapter/ApiAdapter.java [hunk 8]
```diff
@@ -17,12 +17,12 @@
...
 import io.gravitee.definition.model.DefinitionVersion;
-import io.gravitee.rest.api.model.v4.api.ApiEntity;
-import io.gravitee.rest.api.model.v4.api.GenericApiEntity;
-import io.gravitee.rest.api.model.v4.api.NewApiEntity;
-import io.gravitee.rest.api.model.v4.api.UpdateApiEntity;
+import io.gravitee.rest.api.model.v4.api.*;
 import java.io.IOException;
+import java.util.Map;
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/adapter/ApiAdapter.java]: 🗑️

**Review by jgiovaresco — APPROVED** [review]

**Pull request merged**